FastText/src/models/custom_layers.jl

```diff
@@ -0,0 +1,296 @@
+"""
+ULMFiT - Custom layers
+
+This file contains the custom layers defined for this model:
+    AWD_LSTM
+    WeightDroppedLSTM
+    VarDrop
+    PooledDense
+"""
+
+import Flux: gate, testmode!, _dropout_kernel
+
+reset_masks!(entity) = nothing
+reset_probability!(entity) = nothing
+
+"""
+    drop_mask(x, p)
+
+Drop mask generator
+
+This function generates dropout mask for given 'x' with `p` probability
+    or
+It can be used to generate the mask by giving the shape of the desired mask and probaility
+"""
+function drop_mask(x, p)
+    y = similar(x, size(x))
+    Flux.rand!(y)
+    y .= Flux._dropout_kernel.(y, p, 1 - p)
+    return y
+end
+
+drop_mask(shape::Tuple, p; type = Float32) = (mask = rand(type, shape...);mask .= _dropout_kernel.(mask, p, 1 - p))
+
+#################### Weight-Dropped LSTM Cell ######################
+"""
+Weight-Dropped LSTM Cell
+
+This is an LSTM layer with dropped weights functionality, that is, DropConnect technique
+
+cite this paper to know about DropConnec:
+http://yann.lecun.com/exdb/publis/pdf/wan-icml-13.pdf
+
+Moreover this also follows the Vartional DropOut citeria, that is,
+the drop mask is remains same for a whole training pass.
+This is done by saving the masks in 'maskWi' and 'maskWh' fields
+"""
+mutable struct WeightDroppedLSTMCell{A, V, S, M}
+    Wi::A
+    Wh::A
+    b::V
+    h::S
+    c::S
+    p::Float64
+    maskWi::M
+    maskWh::M
+    active::Bool
+    state0
+end
+
+function WeightDroppedLSTMCell(in::Integer, out::Integer, p::Float64=0.0;
+    init = Flux.glorot_uniform)
+    @assert 0 ≤ p ≤ 1
+    cell = WeightDroppedLSTMCell(
+        init(out*4, in),
+        init(out*4, out),
+        init(out*4),
+        reshape(zeros(Float32, out), out, 1),
+        reshape(zeros(Float32, out), out, 1),
+        p,
+        drop_mask((out*4, in), p),
+        drop_mask((out*4, out), p),
+        true,
+        (Flux.zeros32(out, 1), Flux.zeros32(out, 1))
+    )
+    cell.b[gate(out, 2)] .= 1
```

Comment:
Why does it only mask part of the bias @Chandu-4444?

Reply:
Since it is ported from TextModels.jl I'm not completely sure of this. But I'm sure this would be of the same as this https://github.com/FluxML/Flux.jl/blob/a377cb0637557b4c189fcc904bec65df9a124eb0/src/layers/recurrent.jl#L292.